### DISCUSSION TITLE
Frustrated trying to run the UI of Opensnitch

### DISCUSSION THREAD

**Discussion by Rainy-Window** [evilsocket/opensnitch#1044] [General]
I've read and tried everything that was in the issues section related to GUI and bugs. However, it didn't work for me..
the daemon is running in the background, i tried `opensnitchd -check-requirements` and they're all there.  

i also tried to apply the fixes mentioned here: [https://github.com/evilsocket/opensnitch/wiki/GUI-known-problems]
but nothing worked for me. 

i'm trying to install those versions:
opensnitch_1.6.2-1_amd64.deb
python3-opensnitch-ui_1.6.3-1_all.deb

OS: mint 21.1
Kernel: 5.15.0-83-generic 

opensnitch was working for me for about 5 months but with this version `opensnitch (1.6.0-rc.4-1)`
suddenly it stopped working and nothing boots on startup. i tried several times to reboot and run the UI but nothing worked. 
so i decided to remove it and install the latest mentioned version `1.6.2-1` and i followed every step on the main page, but still facing the same problem. i even tried to revert and install the older version and still it didn't work ..

this is the most recent error i get when i try to run `sudo opensnitch-ui` 

<details><summary>Details</summary>
<p>

Traceback (most recent call last):
  File "/usr/bin/opensnitch-ui", line 36, in <module>
    import grpc
ModuleNotFoundError: No module named 'grpc'
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 153, in apport_excepthook
    with os.fdopen(os.open(pr_filename,
FileNotFoundError: [Errno 2] No such file or directory: '/var/crash/_usr_bin_opensnitch-ui.0.crash'

Original exception was:
Traceback (most recent call last):
  File "/usr/bin/opensnitch-ui", line 36, in <module>
    import grpc
ModuleNotFoundError: No module named 'grpc'

</p>
</details> 

i've had different errors before.. But after multiple removing packages and downgrading some of them and reinstalling others (based on the solutions mentioned in the other issues) .. This is what i got in the end.

EDIT:
tried this command `sudo apt install python3-grpcio python3-protobuf python3-slugi` 
and then `sudo opensnitch-ui`

it produced the same error i used to get before that one.. 

<details><summary>Details</summary>
<p>

Themes not available. Install qt-material if you want to change GUI's appearance: pip3 install qt-material.
	 ~ OpenSnitch GUI - 1.6.3 ~
	protobuf: 3.8.0 - grpc: 1.30.2
-------------------------------------------------- 

QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
gRPC Max Message Length: None
                  Bytes: 4194304
is new file, or IN MEMORY, setting initial schema version
setting schema version to: 3
setting schema version to: 3
Setting journal_mode:  OFF
Setting DB memory optimizations
schema version: 3
db schema is up to date
Loading translations: /usr/lib/python3/dist-packages/opensnitch/utils/../i18n locale: en_US
using IPASN DB: /usr/lib/python3/dist-packages/data/ipasn_20140513_v12.dat.gz

</p>
</details> 

and it always freezes at that line..

**Comment by gustavo-iniguez-goya** [comment]
Hi @ghostiy ,

I'm already sorry that you're experiencing so many problems. This seems to be caused by Ubuntu/Mint's python3-grpcio package as explained here: #647 (and https://bugs.launchpad.net/ubuntu/+source/grpc/+bug/1971114)
Let's see if I can help you.

Firstly, don't run the GUI as root. It's not necessary, and it's better to run it with a non-privileged user.
Secondly, post the output of the following commands to know what grpcio versions do you have installed:
`~ $ pip list installed | grep -E "(protobuf | grpc)"`
`~ $ sudo pip list installed | grep -E "(protobuf | grpc)"`
`~ $ dpkg -l | grep -E "(protobuf | grpc)"`

In my case as [explained here](https://github.com/evilsocket/opensnitch/discussions/1003#discussioncomment-6642001), I only have installed the following grpcio packages:
```
~ $ dpkg -l | grep grpcio
ii  python3-grpcio                   1.30.2-3build6
~ $ pip3 list installed | grep grpc
grpcio                               1.44.0
~ $ sudo pip3 list installed | grep grpc
grpcio                               1.30.2
```

protobuf 3.12.4.

**Reply by Rainy-Window** [reply]
Hello @gustavo-iniguez-goya 

Thank you for your fast response
I'm confused for what version to install for what to make it work for me. 
other issues mentioned different version, but i don't understand why it's not working for me till now. 

However, my output of the mentioned commands is like this :

> protobuf               4.21.12
protobuf               3.8.0
ii  python3-protobuf                           3.12.4-1ubuntu7.22.04.1                     amd64        Python 3 bindings for protocol buffers

also for my grpcio packages (they seem to be little difference than yours):

> ii  python3-grpcio                             1.30.2-3build6                              amd64        GRPC system (Python 3)
grpcio                 1.58.0
grpcio                 1.30.2

and this is a screenshot of the running processes of opensnitch:
(please notice that the `sudo opensnitch-ui` command still running in my bash for 2 hours now - i didn't close the bash window)

![Screenshot at 2023](https://github.com/evilsocket/opensnitch/assets/30316475/ac44b369-89a9-48e9-b7b1-aea5f5d579bd)

**Reply by gustavo-iniguez-goya** [reply]
if it's still running kill the process with: `~ $ sudo kill -9 6948 6949 6950`

The problem is the grpcio version: `grpcio 1.30.2`

If you launch the GUI as your user (without sudo) it should show up on the systray.
If you still want to launch it with sudo, then you need to install grpcio==1.44.0 system-wide: 
`~ $ sudo pip3 install grpcio==1.44.0`

**Reply by Rainy-Window** [reply]
this is what comes out when i try to type `opensnitch` without `sudo`

```
Traceback (most recent call last):
  File "/usr/bin/opensnitch-ui", line 42, in <module>
    from opensnitch.service import UIService
  File "/usr/lib/python3/dist-packages/opensnitch/service.py", line 15, in <module>
    from opensnitch import ui_pb2
  File "/usr/lib/python3/dist-packages/opensnitch/ui_pb2.py", line 32, in <module>
    _descriptor.EnumValueDescriptor(
  File "/home/ver/.local/lib/python3.10/site-packages/google/protobuf/descriptor.py", line 755, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 153, in apport_excepthook
    with os.fdopen(os.open(pr_filename,
FileNotFoundError: [Errno 2] No such file or directory: '/var/crash/_usr_bin_opensnitch-ui.1000.crash'

Original exception was:
Traceback (most recent call last):
  File "/usr/bin/opensnitch-ui", line 42, in <module>
    from opensnitch.service import UIService
  File "/usr/lib/python3/dist-packages/opensnitch/service.py", line 15, in <module>
    from opensnitch import ui_pb2
  File "/usr/lib/python3/dist-packages/opensnitch/ui_pb2.py", line 32, in <module>
    _descriptor.EnumValueDescriptor(
  File "/home/ver/.local/lib/python3.10/site-packages/google/protobuf/descriptor.py", line 755, in __new__
    _message.Message._CheckCalledFromGeneratedFile()
TypeError: Descriptors cannot not be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates

```

**Reply by gustavo-iniguez-goya** [reply]
for this error execute without sudo: `~ $ pip3 uninstall protobuf`
https://github.com/evilsocket/opensnitch/issues/741#issuecomment-1523369154

is part of the same problem.

**Reply by Rainy-Window** [reply]
@gustavo-iniguez-goya 

it works!!!!!!!
i've did this step before.. But i don't know why it gave an error …maybe there was something wrong with grpcio that prevents opensnitch from running..
thank you so much for your effort & time 
thank you <3

**Reply by gustavo-iniguez-goya** [reply]
:partying_face: :)
